neuralpredictors/training/tracking.py

```diff
@@ -146,3 +153,360 @@ def asdict(self, time_key="time", make_copy=True):
         log_copy = {k: np.copy(v) if make_copy else v for k, v in self.log.items()}
         log_copy[time_key] = np.copy(self.time) if make_copy else self.time
         return log_copy
+
+
+class AdvancedTracker(Tracker):
+    """
+    This class implements a more advanced, universal tracker that offers many useful features:
+     - Store logging information in an arbitrary hierarchical structure, e.g.:
+       ```
+       {
+        "learning_rate" : [0.1, 0.08],
+        "Training":
+        {
+            "img_classification": {
+               "accuracy" : [345.234234, 242.34],
+               "loss" : [0.342342, 0.78432],
+               "normalization": [4,3],
+            }
+            "neural_prediction": {
+               "accuracy" : [345.234234, 242.34],
+               "loss" : [0.342342, 0.78432],
+               "normalization": [4,3],
+            }
+        }
+        "Validation": {
+             "img_classification": {
+               "accuracy" : [348.234234, 0.0],
+               "loss" : [0.42342, 0.0],
+               "normalization": [2,1],
+             },
+             "patience": [0],
+         },
+       }
+       ```
+     - Retrieving and manipulating information via a simple hierarchical key,
+       e.g. `("Training","img_classification","loss")`
+     - Scores for the same epoch are automatically accumulated via `log_objective(...)` (usually unnormalized)
+     - Scores can be automatically normalized if `"normalization"` (e.g. total examples) is tracked on the same hierarchy level
+     - Automatic per-epoch initialization for each objective when a new epoch is started via `start_epoch()`
+     - Automatic output of current logs to a logger or even tqdm
+     - Easy extraction and reloading of tracker state via `state_dict()`, `from_dict(sd)` and `load_state_dict(sd)`
+
+
+     Example usage:
+     1. Specify the log structure and initial values:
+         ```
+        objectives = {
+            "lr": 0,
+            "training": {
+                "img_classification": {"loss": 0, "accuracy": 0, "normalization": 0}
+            },
+            "validation": {
+                "img_classification": {
+                    "loss": 0,
+                    "accuracy": 0,
+                    "normalization": 0,
+                },
+                "patience": 0,
+            },
+        }
+        tracker = AdvancedTracker(
+            main_objective=("img_classification", "accuracy"), **objectives
```

Comment:
Why is `main_objective` not starting with the `"training"` key?

Reply:
I added a comment about that in the docstring of the init now.